flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullAwareComparator.scala

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.api.table.typeinfo
+
+import org.apache.flink.api.common.typeutils.TypeComparator
+import org.apache.flink.core.memory.{MemorySegment, DataOutputView, DataInputView}
+
+/**
+ * Null-aware comparator that wraps a comparator which does not support null references.
+ */
+class NullAwareComparator[T](
+    val wrappedComparator: TypeComparator[T],
```

Comment:
I think right now, the `extractKeys` and `getFlatComparators()` methods do not work as expected. These methods are used to compare records from different types, i.e., in a Join. For the comparison, the key fields are extracted by `extractKeys` method that corresponds with the input and then compared using the flat comparators. Hence, the number of keys and comparators must always be the same and the comparators must be able to handle null values. If we want to support composite types as keys, we must handle the case of null-valued fields very carefully, i.e., return the right number of null values for each nested field + all nested comparators must support null values.